File: 0007-reverse/python/Solution.py
from typing import List


class Solution:
    def reverse(self, x: int) -> int:
        min_val = -(2**31)
        max_val = 2**31 - 1
        res = 0
        while x != 0:
            if x < 0:
                tmp = x % -10
                x = -(x // -10)
            else:
                tmp = x % 10
                x = x // 10
            res = res * 10 + tmp
            if res > max_val or res < min_val:
                return 0
        return res

    def reverse1(self, x: int) -> int:
        minval = (-2) ** 31
        maxval = 2**31 - 1
        temp = str(x)
        if x < 0:
            return (
                int("-" + temp[:0:-1])
                if minval <= int("-" + temp[:0:-1]) <= maxval
                else 0
            )
        else:
            return int(temp[::-1]) if minval <= int(temp[::-1]) <= maxval else 0


if __name__ == "__main__":
    s = Solution()
    print(s.reverse1(123))
    print(s.reverse1(-123))
    print(s.reverse1(120))
    print(s.reverse1(0))

File: 0007-reverse/cpp/Solution.cpp
/**
 * @file Solution.cpp
 * @author your name (you@domain.com)
 * @brief 
 * @version 0.1
 * @date 2022-03-02
 * @details 7. 整数反转
 * @copyright Copyright (c) 2022
 * 
 */
#include <limits.h>
#include <iostream>

class Solution {
public:
    int reverse(int x) {
        int res = 0;
        while (x) {
            int tmp = x % 10;
            x /= 10;
            if (res > INT_MAX / 10 || (res == INT_MAX / 10 && tmp > 7)) return 0;
            if (res < INT_MIN / 10 || (res == INT_MIN / 10 && tmp < -8)) return 0;
            res = res * 10 + tmp;
        }
        return res;
    }
};

int main(int argc, char const *argv[])
{
    /* code */
    std::cout <<Solution().reverse(123) << std::endl;
    std::cout <<Solution().reverse(-123) << std::endl;
    return 0;
}
